### PR TITLE
Fix false negative ``no-member`` when the code was a `manual` augmented assignments

### LINKED ISSUE #4562: False negative and false positive 'no-member' errors with `+=`
### Steps to reproduce

Given a file `a.py`:

```python
# pylint: disable=missing-module-docstring
# pylint: disable=too-few-public-methods
# pylint: disable=missing-class-docstring
# pylint: disable=invalid-name


class A:
    value: int


obj = A()
obj.value += 1
```

Given a file `b.py`:

```python
# pylint: disable=missing-module-docstring
# pylint: disable=too-few-public-methods
# pylint: disable=missing-class-docstring
# pylint: disable=invalid-name


class A:
    value: int


obj = A()
obj.value = 1 + obj.value
```

Given a file `c.py`:

```python
# pylint: disable=missing-module-docstring
# pylint: disable=too-few-public-methods
# pylint: disable=missing-class-docstring
# pylint: disable=invalid-name


class A:
    value: int


obj = A()
obj.value += 1
obj.value = 1 + obj.value
```

### Current behavior

Results:

```sh
$ pylint a.py b.py c.py
************* Module a
a.py:12:0: E1101: Instance of 'A' has no 'value' member (no-member)

-------------------------------------------------------------------
Your code has been rated at 6.15/10
```

### Expected behavior

There should also be `no-member` issues in `b.py` and `c.py`.

### pylint --version output

Result of `pylint --version` output:

```
pylint 2.8.2
astroid 2.5.6
Python 3.8.10 (default, May 13 2021, 15:41:32) 
[GCC 9.3.0]
```

### PR DESCRIPTION
**Pull request by clavedeluna**
## Type of Changes

|     | Type                   |
| --- | ---------------------- |
| ✓   | :bug: Bug fix          |

## Description

Add code to detect `no-member` when an augmented assign is done "manually", that is, when a user does this

`obj.value = 1 + obj.value`
instead of 
`obj.value += 1`

If `no-member` would be raised in the second case, it should also be raised in the first case, which is why this PR is here!

Closes #4562 (altho I can remove this auto-close if we want to keep it open for the `pydantic` false positive case, though I Personally think it should be a separate issue)

### CODE REVIEW

File: pylint/checkers/utils.py
Comment (clavedeluna) [inline pylint/checkers/utils.py:1998]: this entire util func is pretty ugly ... all ears for improvements

File: pylint/checkers/utils.py
Comment (Pierre-Sassoulas) [inline pylint/checkers/utils.py:1983]: As children is only used in the second list  comprehension, imbricating both list comprehension would be more efficient. 

I'm on mobile so I can't check get_children api but I wouldn't be surprised if you can get all the children of a certain type directly.

File: pylint/checkers/utils.py
Comment (Pierre-Sassoulas) [inline pylint/checkers/utils.py:1984]: We can avoid doing the following if in some cases like this:
```suggestion
    if not binops or binops[0] ! = "+":

File: pylint/checkers/utils.py
Comment (Pierre-Sassoulas) [inline pylint/checkers/utils.py:1998]:
```suggestion
    return attr.attrname == node.attrname
```

File: pylint/checkers/utils.py
Comment (Pierre-Sassoulas) [inline pylint/checkers/utils.py:1989]:
```suggestion

File: pylint/checkers/utils.py
Comment (Pierre-Sassoulas) [inline pylint/checkers/utils.py:1995]:
```suggestion
    else:
```

**Review by Pierre-Sassoulas — CHANGES_REQUESTED** [review]
Looks pretty good already, thank you. I made a small review, there's an issue with the pipeline to fix too.

**Comment by DanielNoord** [comment]
@Pierre-Sassoulas I was thinking: does it make sense to make this a code style check? `consider-using-augmented-assign`. We would get a nice primer result to check for false positives on this util function.

**Comment by Pierre-Sassoulas** [comment]
Sounds good !

**Comment by DanielNoord** [comment]
@clavedeluna I'll have a look at converting this into a message tomorrow. You can do so as well if you want to? I'd be happy to guide you in the right direction?

File: pylint/checkers/utils.py
Comment (clavedeluna) [reply pylint/checkers/utils.py:1989]: I'm much more in favor of 
```
var = None
if ...
   var = 1
elif ...
  var = 2

if not var
```

but hey I"ll take your suggestion

Comment (Pierre-Sassoulas) [reply pylint/checkers/utils.py:1989]: I did not consider the case where var is none from the if, I'm not opinionated about this you asked about possible style fix so I tried :)

File: pylint/checkers/utils.py
Comment (clavedeluna) [reply pylint/checkers/utils.py:1983]: I don't see a way to get children of a certain type https://pylint.pycqa.org/projects/astroid/en/latest/api/astroid.nodes.html#astroid.nodes.AugAssign.get_children
but will def combine the list comps

Comment (Pierre-Sassoulas) [reply pylint/checkers/utils.py]: Maybe node_of_classes but I'm not sure it's only the children that are returned. Sorry for suggesting things that I can't test myself right now, I'm on mobile.

**Comment by clavedeluna** [comment]
whoops... broke the tests

**Comment by clavedeluna** [comment]
Alright so maybe I'm missing something obvious but how is
`pylint/checkers/utils.py:1996: error: Returning Any from function declared to return "bool`
an issue for `return attr.attrname == node.attrname`

is that not always a bool? do I wrap it in bool??

**Comment by DanielNoord** [comment]
> Alright so maybe I'm missing something obvious but how is `pylint/checkers/utils.py:1996: error: Returning Any from function declared to return "bool` an issue for `return attr.attrname == node.attrname`
> 
> is that not always a bool? do I wrap it in bool??

This is because `astroid` is not typed so you're essentially comparing `Any with Any`. If you add a disable for this specific error with `type: ignore` we will remove it after `astroid` does have typing.

**Comment by clavedeluna** [comment]
> we will remove it after `astroid` does have typing.

ah makes sense! I'm getting more and more familiar with the code base so thanks for the tip.

**Review by Pierre-Sassoulas — DISMISSED** [review]
LGTM, thank you @clavedeluna !

**Comment by DanielNoord** [comment]
I'm working a PR that incorporates these changes to get a primer result. Should be done in 30 minutes or something like that.

**Comment by Pierre-Sassoulas** [comment]
I'll let you merge this one @DanielNoord

**Comment by DanielNoord** [comment]
See https://github.com/PyCQA/pylint/pull/7514.

I had to change quite a bit. Preferably we merge that PR first and then use the utility introduced there in this PR. The robustness we get from checking the new checker on the primer is a huge benefit compared to the functional tests we can think of ourselves.

**Comment by clavedeluna** [comment]
> See #7514.
> 
> I had to change quite a bit. Preferably we merge that PR first and then use the utility introduced there in this PR. The robustness we get from checking the new checker on the primer is a huge benefit compared to the functional tests we can think of ourselves.

I will rebase and re-do this PR once that one is merged

**Review by Pierre-Sassoulas — APPROVED** [review]
Thank you for inspiring the work on the augmented assignment checker, and thank you for fixing this !

**Pull request merged**
